### ISSUE TITLE
FSDP 2 + cpu_ram_efficient_loading=True fails for models with buffers (e.g. DS 3.1)

### ISSUE THREAD

**Issue by d5031** [huggingface/accelerate#3898]
err is: `'Tensor' object has no attribute 'device_mesh'` at

        for param_name, sharded_param in meta_sharded_sd.items():
            device_mesh = sharded_param.device_mesh

in `prepare -> _prepare_fsdp2 -> fsdp2_prepare_model -> fsdp2_load_full_state_dict`

for buffers like `base_model.model.model.layers.3.mlp.gate.e_score_correction_bias` defined as

    self.register_buffer("e_score_correction_bias", torch.zeros(self.n_routed_experts))

a very lazy fix (but it works) is:

    if accelerator.is_main_process:
        for param_name, sharded_param in meta_sharded_sd.items():
            full_param = full_sd[param_name].detach()

            if isinstance(sharded_param, DTensor):
                device_mesh = sharded_param.device_mesh
                full_param = full_param.to(device_mesh.device_type)
                if isinstance(full_param, DTensor):
                    full_param = full_param.to_local()
                dist.broadcast(full_param, src=0, group=dist.group.WORLD)

                sharded_tensor = distribute_tensor(full_param, device_mesh, sharded_param.placements)
                to_contiguous, casting_dtype = _infer_parameter_dtype(model, param_name, full_param)
                sharded_tensor = _cast_and_contiguous(sharded_tensor, to_contiguous, casting_dtype)
                sharded_sd[param_name] = sharded_tensor
            else:
                # Buffer/unsharded: broadcast and keep as plain tensor
                full_tensor = full_param.to(accelerator.device)
                if isinstance(full_tensor, DTensor):
                    full_tensor = full_tensor.to_local()
                dist.broadcast(full_tensor, src=0, group=dist.group.WORLD)

                to_contiguous, casting_dtype = _infer_parameter_dtype(model, param_name, full_tensor)
                full_tensor = _cast_and_contiguous(full_tensor, to_contiguous, casting_dtype)
                sharded_sd[param_name] = full_tensor
    else:
        for param_name, sharded_param in meta_sharded_sd.items():
            # allocate receive buffer on local device regardless of DTensor/buffer
            full_tensor = torch.empty(
                sharded_param.size(),
                device=accelerator.device,
                dtype=sharded_param.dtype,
            )
            dist.broadcast(full_tensor, src=0, group=dist.group.WORLD)

            if isinstance(sharded_param, DTensor):
                device_mesh = sharded_param.device_mesh
                sharded_tensor = distribute_tensor(full_tensor, device_mesh, sharded_param.placements)
                to_contiguous, casting_dtype = _infer_parameter_dtype(model, param_name, full_tensor)
                sharded_tensor = _cast_and_contiguous(sharded_tensor, to_contiguous, casting_dtype)
                sharded_sd[param_name] = sharded_tensor
            else:
                to_contiguous, casting_dtype = _infer_parameter_dtype(model, param_name, full_tensor)
                full_tensor = _cast_and_contiguous(full_tensor, to_contiguous, casting_dtype)
                sharded_sd[param_name] = full_tensor

Might be @S1ro1  can have a look? Thanks
